### ISSUE TITLE
Allow shared config dirs configured via .clj-kondo/config.edn

### ISSUE THREAD

**Issue by borkdude** [borkdude/clj-kondo#992]
We should be able to configure extra config paths in `.clj-kondo/config.edn`: a clj-kondo config classpath if you will.

We already allow extra config files via `--config` but this isn't always accessible from different editor tooling and it doesn't help for shared hook code.

E.g. project1 and project2 both use roughly the same config and hook. They should be able to point to a shared config dir  with a `config.edn`  and/or `/hooks/**` directory that will be picked up on.

We could have something like:

``` 
:config-paths ["../base/clj-kondo", "~/dev/clj-kondo/configs/rum/.clj-kondo"]
``` 

All of the config.edn files in these directories will be merged from left to right and these directories will also form the classpath for hook code. 

For personal projects your config-path could just be in your home directory like `["~/.clj-kondo"]`. 

Our local config.edn should always be merged in last as to override the config from config paths.

One possible problem is not exactly having the same paths on CI as you would have locally. Maybe we should ignore non-existing paths silently. 

We will still have *only one cache directory*, the one in .clj-kondo/.cache, since it's not great to have the cache be spread over multiple directories which may have different formats due to different clj-kondo versions being used.

The hooks code from clj-kondo's examples directory can become its own repository, so users can git clone it and add the config (selectively, per library) to their project:

`["~/dev/clj-kondo/configs/rum/.clj-kondo"]`.

Should we expand `~` to the `HOME` ? Should we have conditionals based on environment variables for CI usage? Or is simply ignoring non-existent paths enough? In https://github.com/borkdude/clj-kondo/blob/master/doc/ci-integration.md it is documented that `--config` overrides the local config, which is convenient for CI to disable some things. I guess CI can also choose to replace config paths using the `--config` option + `^:replace`.

TBD:

- interpret  `~`
- transitive config paths, i.e. config paths from a config path's config.edn should also be resolved?
- support .jar files

Related issues:

#559, #736, #790, https://github.com/borkdude/clj-kondo.lsp/pull/9

cc @orestis @kwrooijen @rschmukler @Pancia @Limess @bfontaine @glisom

**Comment by glisom** [comment]
I think this would solve for https://github.com/borkdude/clj-kondo.lsp/pull/9 especially if it just silently ignored missing config-paths.

**Comment by borkdude** [comment]
Here's the idea expressed in a screenshot:

<img width="649" alt="config-paths" src="https://user-images.githubusercontent.com/284934/92307549-96f59b00-ef97-11ea-85e3-1d643f1757f8.png">

**Comment by hoxu** [comment]
It would be nice if there was a solution that just worked ouf of box with default config without having to configure a list of config paths. Perhaps allow disabling such autoconfiguration either completely or for certain packages, in case there are issues.

If everything you need is something as simple as:

```clojure
{:lint-as {mount.core/defstate clojure.core/def}}
```

then having to add config-paths quickly becomes as cumbersome as the config itself. Especially if you would need one config-path for every project that would add something trivial like that.

Personally I wouldn't see it as horrible if clj-kondo included some default lint-as configs for the most popular packages. That would allow you to change the configuration format in the future without having to take into consideration repositories that are not under your control and have different release cycles.

**Comment by borkdude** [comment]
@hoxu You can still put all your favorite config into one config path and point to that one directory. You can check that config into a git repo and clone it on every machine you own.

**Comment by borkdude** [comment]
Eventually the most popular configs can end up in clj-kondo itself, but having it in user space for a while has benefits. I've been making tweaks to the Rum config a couple of times now and it still isn't right (see https://github.com/borkdude/clj-kondo/issues/987). So after it's been incubated for a while in a community repo it can make its way into clj-kondo.

**Comment by borkdude** [comment]
Binaries for playing with this: see branch https://github.com/borkdude/clj-kondo/tree/config-paths. Click on the CircleCI or AppVeyor links in the commit checks. There you can download artifacts. 

Configs repo: https://github.com/clj-kondo/clj-kondo.configs

**Comment by borkdude** [comment]
Binaries for testing this feature:

Linux: https://10893-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2020.07.30-SNAPSHOT-linux-amd64.zip
macOS: https://10892-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2020.07.30-SNAPSHOT-macos-amd64.zip

**Comment by borkdude** [comment]
VSIX (cc @glisom) 

[clj-kondo-2020.7.300.vsix.zip](https://github.com/borkdude/clj-kondo/files/5178698/clj-kondo-2020.7.300.vsix.zip)

**Comment by asafch** [comment]
@borkdude what about working with an abstraction, e.g. package or namespace instead of something very concrete as a file? This is the same approach that [ESlint takes](https://eslint.org/docs/user-guide/getting-started) with the `"extends"` parameter in `.eslintrc.js`.

Example:
`:extends [org.whatever.asach/kondo "1.2.3"]` for a Clojure package and version (like Leiningen's dependencies);
`:extends [asach-kondo.core]` for a Clojure namespace that should be somewhere in the class path (notice it's a symbol and not a string) and include a pre-defined var, maybe `(def config '{})`?

This can solve the problem of making sure that files sit in the directories you expect them to be across local machine, docker image, cloud machine or something else, by just declaring a dependency and having `cli-kondo` work with it.

**Comment by borkdude** [comment]
@asafch One use case for the above is multi-repos.

/work/project1
/work/project2
/work/shared-company-config

So project1 and project2 will refer to shared-config using :config-paths.

Being able to refer to some shared library that you happened to clone just falls out of this.
Adding dependencies to this, can be a second step.

We could additionally allow dependencies that will be looked up in your maven repo or .gitlibs, but then clj-kondo would have to do dependency resolution (via clojure or lein) which will slow linting down - or do this only once and then look in .m2 / .gitlibs?

**Comment by asafch** [comment]
@borkdude indeed, any such resolution would slow down linting, but without it I fear this feature wouldn't bring much value for large teams:

Assuming that `.clj-kondo/config.edn` is part of the project repository, and that we provide only the `:config-paths` option, it creates the burden of having to make sure that the base config file is located in the same directory across all deployment scenarios: a developer's laptop, a CI node, a deployment node, k8s, etc. It can be a real problem, and sometimes impossible due to directory access policies in each scenario (i.e. I can't guarantee that `chmod` behaves the same in each scenario with regards to the given directory). I could get around this by providing a list of config files, where each one targets a different scenario, but this introduces ambiguity, and the off-chance that more than one file would actually be resolved and merged into the end config.

If you would allow describing a dependency based solely on project name and version (let's drop the namespace option), you could optimize the dependency resolution and perform it only once and cache the result under the assumption that the contents of a package won't change when its version is set. The only exception is a `-SNAPSHOT` version - you would have to test for changes in the package each time you lint, e.g. `org.whatever.asach.kondo-1.2.3-20200811.075605-6.jar` has been replaced by `org.whatever.asach.kondo-1.2.3-20200812.180000-2.jar` in `~/.m2/repository/org.whatever.asach.kondo/org.whatever.asach.kondo/1.2.3-SNAPSHOT/`.

**Comment by borkdude** [comment]
@asafch Thanks for the feedback, I do appreciate it, but I also disagree with some of what you said.  Always having to put your shared config in an artifact would be too limiting in my opinion.

In my own team we have a repo like this:

``` 
/repo/.git
/repo/project1/.clj-kondo/config.edn
/repo/project2/.clj-kondo/config.edn
``` 

With the feature as it is now, this can become:

``` 
/repo/.git
/repo/project1/.clj-kondo/config.edn
/repo/project2/.clj-kondo/config.edn
/repo/clj-kondo-shared/config.edn
``` 

This is already a win for us and the usage of `:config-paths ["../../clj-kondo-shared"]` is 100% predictable in our repo. 
The feature as it is now has also has been requested by @glisom in borkdude/clj-kondo.lsp#9. I've repeatedly have had the request to let users put config for their personal projects in `~/.clj-kondo` which is also possible using the feature as it is now, if we are going to interpret tilde (which it doesn't yet, but we may - I'm waiting on some feedback from @puredanger, as he doesn't expand this in deps.edn for some reason).

We could extend `:config-paths` with support for `.jar` files. Together with git libs you can then use this with dependencies via lein or clojure, but pointing to your m2 or .gitlibs dir.  I realize that this is too low level, but it may become a foundation for something more ergonomic where some invocation of your favorite build tool can create these paths for you.

Awaiting more feedback.

**Comment by borkdude** [comment]
What about this:

``` 
{:extends {:paths ["../../shared-configs"]
           :deps {clj-kondo/clj-kondo.configs {:git/url "..."
                                               :sha "..."}
                  org.acme/clj-kondo.configs {:mvn/version "..."}}
           :mvn-repo "~/foobar/.m2" ;; defaults to "~/.m2/repository"
           :gitlibs "~/foobar/.gitlibs/libs" ;; defaults to "~/.gitlibs/libs"
           }}
``` 

So `:paths` is what it is now. But `:deps` will allow you to specify dep coordinates.
You will have to put these deps in your `project.clj`, `build.boot` or `deps.edn` in order for them to be downloaded, that is the task of your build tool. Clj-kondo will just look in the right places based on the config, but won't attempt to download them (since it's not a build tool).

**Comment by robert-stuttaford** [comment]
Particularly keen for a `~/.clj-kondo/config.edn`!

**Comment by johnchristopherjones** [comment]
I’d really appreciate following the XDG spec on this for broader consistency. For user config that would typically mean`.config/clj-kondo/config.edn`, though it’s really `$XDG_CONFIG_HOME/clj-kondo/config.edn`

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

https://unix.stackexchange.com/questions/24347/why-do-some-applications-use-config-appname-for-their-config-data-while-other

**Comment by borkdude** [comment]
If we would resolve tilde (`~`) in paths you could put your home config anywhere you would like using `:config-paths` like above.

So far I've resisted automatically resolving config in home dirs for these reasons:

1) Clj-kondo config belongs to a project and forms an agreement among team members what rules they want to use. 

But I guess linting in CI will be the ultimate agreement and you can enable extra things on your own system if you'd like.

2) If people would put config inside `~/.clj-kondo/config.edn` and would have no `.clj-kondo` in their project yet, there would be a global cache populated in `~/.clj-kondo/.cache` for all edited clojure files on your system, without another `.clj-kondo` directory in the directory hierarchy.

This can be mitigated by choosing `~/.config/clj-kondo` as the default clj-kondo home dir config location, unless `XDG_CONFIG_HOME` indicates otherwise.

If we automatically resolve home dir configs, maybe interpreting tilde on `:config-paths` isn't needed anymore. Alex Miller explained that deps.edn does not interpret tilde because other tooling should also be able to read deps.edn and "The less they have to know, the better", i.e. just get data and don't do magic processing on it.

**Comment by borkdude** [comment]
OK, here is my updated proposal. 

- clj-kondo will resolve config and hook code from `$HOME/.config/clj-kondo` and respect `XDG_CONFIG_HOME` 

Additionally it is possible to extend from configs elsewhere on the filesystem:

```
{:extends
 {:configs
  [{:dep clj-kondo/clj-kondo.configs
    :git/url "https://github.com/clj-kondo/clj-kondo.configs"
    :sha "db781c287c9cfdcfe28610b4bd491dba47ee2af9"
    :root "rum/.clj-kondo"}
   {:dep org.acme/clj-kondo.configs
    :mvn/version "..."}
   {:dir "../../shared-configs"}]
  :mvn/repo "~/foobar/.m2/repository" ;; defaults to "~/.m2/repository"
  }}
``` 

- `:dep` indicates a dependencies' group and library name. You can have `:git/url` and `:mvn/version` like in `deps.edn`. Clj-kondo will search in your `$HOME/.m2/repository` unless otherwise indicated by `:mvn/repo` and `$HOME/.gitlibs`. It will not download these deps for you: your build tool should do this via `project.clj`, `deps.edn`, etc. Additionally you can specify a `:root`, if clj-kondo should look elsewhere than `/` for config. This allows config to be packaged along with normal libraries.

- `:dir` indicates a directory to read config from. Can be absolute or relative.

**Comment by lread** [comment]
If I understand all of the above correctly, a clj-kondo config will be automatically applied if it is found in the expected spot under `$HOME`?

As to your idea:
> But I guess linting in CI will be the ultimate agreement and you can enable extra things on your own system if you'd like.

I dunno.  I feel that project owners have their own lint preferences and contributors should be guided to follow those preferences rather than have their own preferences automatically mixed in.

Perhaps it doesn't matter when the project owner is lint-aloof, but for project owners who are lint-picky, is there to be a facility for a project's lint config to state it wants its and only its lint configuration to apply?

**Comment by borkdude** [comment]
@lread Yes. You can use `^:replace` metadata on your config which will ignore all previous merged configs.

**Comment by borkdude** [comment]
Also you can override everything you want in the local config.edn of course.

**Comment by lread** [comment]
Thanks @borkdude, I think `^:replace` should do the trick for me!

My CI lint script (which I will also encourage contributors to run) does have a bit of `--config` on my `clj-kondo` command command line. But since my script is written in babashka I will easily be able to to have the script merge that config with my project's local config and employ `^:replace` on the resulting config.

**Comment by borkdude** [comment]
@lread The order of merge will be this:

0. clj-kondo's default config (which determines which linters are on and off by default)
1. `~/config/.clj-kondo` 
2. `:config-paths` 
3. local `.clj-kondo/config.edn` 
4. configs provided via `--config`.

At any stage and at any level you can employ ` ^:replace`. E.g. `^:replace {:linters {}}` in your local config.edn will ignore the home config, but also clj-kondo's default config. Thus it will disable all linters. If you use exactly this in step 4, then whatever config was in step 3 will be totally ignored.

I realize now that when having a `config.edn` with:

``` 
^:replace {:config-paths ["foobar"]}
``` 

will undo all config from config-paths.

I think for ignoring only home we can support this:

``` 
{:config-paths ^:replace ["foobar"]}
``` 

because in the default config `:config-paths` contains only the home directory, which is then ignored.

**Comment by hoxu** [comment]
I'm not sure merging home config by default is a good idea if there is a project-specific config.

What are the use cases for home configs?

**Comment by borkdude** [comment]
@hoxu See #736 and #790.

**Comment by Limess** [comment]
I wouldn't expect to merge a home config, only to use it as a fallback if nothing else exists - so as to keep any shared config consistent.

The issue I raised (https://github.com/borkdude/clj-kondo/issues/736) was around a fallback for repos entirely missing clj-kondo configs, but allowing me to use a sensible default.

**Comment by borkdude** [comment]
Not merging home dir config can be opted out with using:

``` 
:config-paths ^:replace []
``` 

so project owners can decide if they want contributors to see their personal lint preferences appear or not.

**Comment by hoxu** [comment]
There is no doubt that you have designed a flexible system that can cover many usecases. But I still have mixed feelings about the "merge home config when project config is present" default.

At the very least, I think documentation examples could include `^:replace` by default.

Will you create a repository for the community to share configuration?

**Comment by rschmukler** [comment]
@borkdude this seems great. Given that you are planning on ignoring unresolved deps, I wonder if it’s worth just automatically including all paths in the deps.edn. The advantage I see is that it’s one less set of deps to have to manage. I could imagine it being a confusing process tracking down that a macro changed between library versions but clj-kondo’s referencing an out dated hook via its coordinates. The downside here is of the performance overhead of looking for configuration files in all those paths, but I wonder if the overhead is actually noticeable - and if it is, could it be cached?

**Comment by borkdude** [comment]
@hoxy Yes, that option will be documented, so it will be clear how to opt out. I think config in home directories is expected to be merged instead of totally overriden, given what other tools like `clojure` do. Linters in other languages might also do that. Anyway, I can see the use case for a company wanting to have all kinds of hooks in your home folder that are used in most projects but being able to add to them project-locally. So if we offer a way to opt out, I think we can make everyone happy.

**Comment by borkdude** [comment]
@hoxy

> Will you create a repository for the community to share configuration?

Yes: https://github.com/clj-kondo/clj-kondo.configs

**Comment by borkdude** [comment]
@rschmukler Interesting thoughts. I might release this with only support for directories and hold off on the deps stuff until later, so we can give it more time. I'll make a different issue once the initial feature is released.

**Comment by borkdude** [comment]
@rschmukler One other way I can see this developing is that each build tool will have their own plugin, which scans the classpath for clj-kondo configs and spits them out to local directories in your `.clj-kondo` dir, e.g. `.clj-kondo/rum/config.edn`, `.clj-kondo/rum/hooks` and the user just has to add `{:config-paths ["rum"]}` to their `config.edn` after running that plugin. This will ensure that the rum config will always be in sync with your dependencies and will also let the build tool do the work, instead of pushing this complexity into clj-kondo.

I'm doing a similar thing here: https://gist.github.com/borkdude/c0987707ed0a1de0ab3a4c27c3affb03
That program scans clojure.spec specs for type info and spits them out to a directory which is included via `:config-paths`.

**Comment by borkdude** [comment]
This will be part of the upcoming config docs introduction:

``` 
Clj-kondo can be configured in four ways, by providing:

- home dir config in `~/.config/clj-kondo/config.edn` (respects `XDG_CONFIG_HOME`)
- project config: a `config.edn` file in the `.clj-kondo` directory (see
  [project setup](../README.md#project-setup))
- `:config-paths` in project `config.edn`: a list of directories that provide additional config
- a command line `--config` file or EDN arguments
- namespace local config using `:clj-kondo/config` metadata in the namespace form

The configurations are merged in the following order, where a later config overrides an earlier config:

- home dir config
- `:config-paths` in project config
- project config: `config.edn`
- command line `--config` args
- namespace local config

The `^:replace` metadata hint can be used to replace parts or all of the
configuration instead of merging with previous ones. The home dir config is
implicitly part of `:config-paths`. To opt out of merging with home dir config
use `:config-paths ^:replace []` in your project config.


Note that namespace local config must always be quoted: `{:clj-kondo/config
'{:linters ...}}` and quotes should not appear inside the config.
```

**Comment by borkdude** [comment]
@rschmukler Here's an idea:

https://github.com/clj-kondo/clj-kondo.configs#installation

Users of deps.edn can add that library to their projects under an alias, then invoke the main, add something to their `:config-paths` and it works.

**Comment by borkdude** [comment]
We could have something similar for lein. Also we could have something similar that searches within the classpath for configs and spits them out.

**Comment by borkdude** [comment]
If we adopt some kind of convention in the `.clj-kondo` dir respecting `.clj-kondo/configs/rum/config.edn`, `.clj-kondo/configs/slingshot/config.edn`, etc. then several tools can emit their own `.clj-kondo/configs/<name>/config.edn` and that will be automatically merged in. The name has to be something unique. 

Additionally, https://github.com/clj-kondo/clj-kondo.configs could scan the classpath for `.clj-kondo` directories, which automatically get spitted into the project .clj-kondo, using that convention. 
E.g. if library.jar has on their classpath `/foo/bar/.clj-kondo/config.edn` we can then spit this out into `$project/.clj-kondo/foo.bar/config.edn`.

Then we would have one tool that can accommodate community config as well as scan libraries for config and bundle them in a project config dir.

**Comment by johnchristopherjones** [comment]
@borkdude I feel really out of my depth with some of the use-cases you're trying to support and the project history, so please pardon me if I just sound lost.  Please ignore if this is somehow just too off-base.

Would it makes sense to piggy-back entirely on `deps.edn`'s `:aliases` and maybe stick a `:clj-kondo/config` in there.  It already has a very similar defined [resolution order](https://clojure.org/reference/deps_and_cli#_deps_edn), respects `$XDG_CONFIG_HOME`, and would `merge-with merge` project and user configs.  It already handles deps and extending deps through aliases.  It also has `:paths`, which seems like it might cover `:config-paths` nicely for "a tool that scans your classpath for library specific clj-kondo configs".  Would this be an appropriate use of `:aliases`?  Would this conflict with other tooling like `lein` and `boot`?  Or maybe is this all just outside clj-kondo's realm because its a distinct interpreter and other things also need do things I'm not considering?

It *seems* simpler—one configuration resolution system for Clojure.  If I understand correctly, the point of of the `{:extends ...}` data structure is that we are, literally, extending the project configuration—just like aliases.  We also have `:paths`.  Clojure and `deps.edn` already have their buckets of surprises for polyglot codebases, so one fewer possibly-subtley-different resolution systems to worry about seems like a win, at least superficially.

Putting a spin on the idea, maybe `deps.edn` is inappropriate for some reason, but perhaps `tools.deps` can be purposed to resolve clj-kondo's configuration with its own [filesystem/env definitions](https://clojure.org/reference/deps_and_cli#_directories)?  In that case, the resolution rules remain very familiar to deps.edn users but the specific files and paths are different.

I'm just musing here on where the local minima is for the principle of least surprise.  I'm probably really ignorant about a lot of major concerns here.  Your considerations here seem really thoughtful, thorough, the solution seems extensible, and you've cleared thought about this problem a hell of a lot more than I have.

**Comment by borkdude** [comment]
@johnchristopherjones Short answer: no :).

A couple of reasons: clj-kondo should not be coupled with deps.edn only and must be able to work independently from build tools (leiningen, maven, boot, tools.deps). Deps.edn is still alpha (see recent discussions in Clojurians slack). Clj-kondo also has different deep-merge semantics than deps.edn (e.g. it concats vectors). I don't want to depend on implementations in tools.deps and hope they work out alright for clj-kondo. There's also a reason clj-kondo has a config directory rather than just one file, which is used for storing analysis cache. Using tools.deps in clj-kondo won't work, since you cannot compile that with GraalVM and will also slow down things.

**Comment by borkdude** [comment]
Re: [this](https://github.com/borkdude/clj-kondo/issues/992#issuecomment-689026962) idea, I think I'll go for something explicit instead of automatically including things in the config from paths inside `.clj-kondo`:

``` 
$ clojure -A:clj-kondo/configs --lib rum
Removing previous rum config
Copying rum config
Add "configs/rum" to :config-paths in .clj-kondo/config.edn to activate configs.
```

**Comment by borkdude** [comment]
The above solution doesn't require any changes to the current implementation on master. Resolving configs from deps can be implemented as part of this tool: https://github.com/clj-kondo/config. So I think the last update to the README covers everything https://github.com/borkdude/clj-kondo/issues/992#issuecomment-688731065 that's needed and the complexity of dealing with dependencies and build tools can be managed in clj-kondo/config.

**Comment by SevereOverfl0w** [comment]
Given you're dealing with config in multiple operating systems, mentioning $HOME at all makes no sense.

Windows has it's own location for config. Although many tools get this wrong, Mac does as well.

https://github.com/dirs-dev/directories-jvm/blob/main/README.md is the correct way to get a config directory for your tool.

**Comment by borkdude** [comment]
@SevereOverfl0w Does tools.deps respect all of this as well? On my Mac there already many tools listing their configs in `~/.configs`. On Windows, WSL expects config in `C:\Users\foobar\.wslconfig`. So there doesn't seem to be one "the correct way"?

**Comment by borkdude** [comment]
@SevereOverfl0w Another thought:

Maybe there's a difference between config that a GUI app stores somewhere and locations where users are going to edit their config manually. Personally I would find `%APPDATA%/guid-something-something/....` hard to find. Using `%HOME\.config\clj-kondo` everywhere seems like a more predicable and user-friendly location.

**Comment by borkdude** [comment]
Interesting precedent:

Scoop (popular Windows package manager comparable to brew) does the same thing as clj-kondo is going to do on Windows:

https://github.com/lukesampson/scoop/blob/master/lib/core.ps1#L967-L968

It also respects `XDG_CONFIG_HOME`. I think clj-kondo is in good company there.

**Comment by borkdude** [comment]
If people using roaming profiles on Windows want to change their config location:

<img width="609" alt="Screenshot 2020-09-09 at 14 21 00" src="https://user-images.githubusercontent.com/284934/92597753-20fb7780-f2a8-11ea-9956-ec725869dd5c.png">

<img width="428" alt="Screenshot 2020-09-09 at 14 22 28" src="https://user-images.githubusercontent.com/284934/92597772-25c02b80-f2a8-11ea-9a8b-d5890546bebb.png">

**Comment by rschmukler** [comment]
@borkdude thanks for all your thought and the proof of concept with the `clj-kondo/config` alias. Some thoughts as I imagine using it:

1) Are the files copied from the version resolved by the deps.edn? Doesn't that have the same problem of libraries being able to get out of sync with the hooks installed? Sure, now they exist as stale config files instead of stale `deps.edn` coordinates, but I'm not sure it's too different.

2) Would you envision projects checking in these hook config artifacts into version control? Or would you imagine ignoring `.clj-kondo/configs` in gitignore?

What about adding a key to `.clj-kondo/config.edn` of `{:resolve-configs [teknql/systemic]}` where the configs are a vector of symbols that correspond to entries in a `deps.edn` file. This solves the versions potentially getting out of date, as well as requiring explicit opt-in. The other nice behavior here is that you can warn if there aren't any hooks associated with the requested symbol. With this approach, it'd probably also be useful to specify an alias in case there are `extra-deps`. Shadow-cljs lets you supply a vector of aliases and merges all of them, and I've found that to be useful in my own projects.

Happy to move this discussion elsewhere if you prefer.

**Comment by borkdude** [comment]
@rschmukler Good questions. I've also been thinking about this and my current thinking is that this can become part of `clj-kondo/config`: https://github.com/clj-kondo/config/issues/1.

Whether users want to check in their `.clj-kondo/configs` is up to them, I don't see a problem with it.

**Comment by borkdude** [comment]
So let's continue discussing in https://github.com/clj-kondo/config/issues/1.
